### ISSUE TITLE
RFC: Task extending/variants

### ISSUE THREAD

**Issue by milesj** [moonrepo/moon#849]
A common practice in the JavaScript ecosystem is to have an npm script that runs another npm script while passing additional arguments/options to it. A popular example is linting:

```json
{
  "scripts": {
    "lint": "eslint .",
    "lint:fix": "npm run lint --fix"
  }
}
```

This is extremely easy in Node.js but is quite difficult in moon, and is currently not supported. npm scripts are simply just strings, arbitrary ones at that, which are then run in a shell. moon tasks are more akin to structured models that represent a single command (with args), as we need to take inputs, outputs, options, merging, and other factors into account.

The only current solution to this problem in moon is to duplicate the task. This can be somewhat ideal because the extending task may require different inputs/options/etc. 

```yaml
tasks:
  lint:
    command: 'eslint .'
    inputs:
      - '@globs(sources)'
      - '@globs(tests)'
      - '*.js'
      - '.eslintrc.js'
      - 'tsconfig.json'
      - '/.eslintignore'
      - '/.eslintrc.js'
      - '/tsconfig.eslint.json'
      - '/tsconfig.options.json'
  lint-fix:
    command: 'eslint . --fix'
    local: true
    inputs:
      - '@globs(sources)'
      - '@globs(tests)'
      - '*.js'
      - '.eslintrc.js'
      - 'tsconfig.json'
      - '/.eslintignore'
      - '/.eslintrc.js'
      - '/tsconfig.eslint.json'
      - '/tsconfig.options.json'
```

With that being said, we’d love to support task extending/variants as a first-class feature in moon, but… we cannot decide on an API. We’d love your input on this matter:

## Other examples

Before we dive further into the RFC, we want to provide some additional examples that we believe would be solved by this RFC, and ones you may have encountered.

- For Rust projects, building debug and release modes.

```yaml
tasks:
  build:
    command: 'cargo build'
    inputs:
      - 'crates/*/src/**/*'
      - 'crates/*/benches/**/*'
      - 'crates/*/tests/**/*'
      - 'crates/*/Cargo.toml'
      - 'Cargo.toml'
    outputs:
      - 'target/debug/moon'

  build-release:
    command: 'cargo build --release'
    inputs:
      - 'crates/*/src/**/*'
      - 'crates/*/benches/**/*'
      - 'crates/*/tests/**/*'
      - 'crates/*/Cargo.toml'
      - 'Cargo.toml'
    outputs:
      - 'target/release/moon'
    options:
      runInCI: false
```

- For Babel, building CJS/ESM/etc.

```yaml
tasks:
  build:
    command: 'babel ./src -o ./lib'
    env:
      FORMAT: 'cjs'
    # ...

  build-esm:
    command: 'babel ./src -o ./esm'
    env:
      FORMAT: 'esm'
    # ...
```

- For Jest, running tests and updating snapshots.

```yaml
tasks:
  test:
    command: 'jest .'
    # ...

  test-update:
    command: 'jest . -u'
    # ...
```

- For Prettier, checking vs writing.

```yaml
tasks:
  format:
    command: 'prettier . --check'
    # ...

  format-write:
    command: 'prettier . --write'
    # ...
```

## Requirements & goals

The following items *must* be met to achieve full functionality:

- Variants are a distinct task and will be ran in CI automatically (if `options.runInCI` is true) alongside the base task.
- Variants _cannot_ override the `command` but can define additional `args`.
- Variants _must_ be able to define `options`.
    - Merging strategies will be applied.
- The base task _must_ be a valid command on its own, and should be runnable, even if a no-op.

## Option 1) New `variants` field in task

The first option is to support a `variants` field within a task’s config. This field would be a map, where the key is the variant name, and the value is a nested task config. 

The nested task config can define all task settings that the base task would (inputs, outputs, options, etc), except for `command`. To modify arguments, use the `args` setting instead. Merge strategies will be applied when merging the base and variant tasks.

```yaml
tasks:
  build:
    command: 'cargo build'
    inputs:
      - 'crates/*/src/**/*'
      - 'crates/*/benches/**/*'
      - 'crates/*/tests/**/*'
      - 'crates/*/Cargo.toml'
      - 'Cargo.toml'
    outputs:
      - 'target/debug/moon'
    variants:
      release:
        args: '--release'
        outputs:
          - 'target/release/moon'
        options:
          mergeOutputs: 'replace'
          runInCI: false
        
  lint:
    command: 'eslint .'
    inputs:
      - '@globs(sources)'
      - '@globs(tests)'
      - '*.js'
      - '.eslintrc.js'
      - 'tsconfig.json'
      - '/.eslintignore'
      - '/.eslintrc.js'
      - '/tsconfig.eslint.json'
      - '/tsconfig.options.json'
    variants:
      fix:
        args: '--fix'
        local: true
```

With this option, we’d also need a way to target these tasks, and as such, we’d support a trailing variant on the target ID, `scope:task:variant`. For example, the above targets would be `project:build:release` and `project:lint:fix` respectively.

### Pros

- The variants are co-located within the task itself. No indirection.
- Including the variant in the target ID is a nice experience.
- The terminology variants is a very well understood concept.
- Large settings (like `inputs`) can be entirely omitted for variant tasks.

### Cons

- The YAML nesting is getting too deep.
- Task inheritance will be very complicated…
- Poor readability, even in that small example above. Now imagine a giant task with 5 variants.
- Loading these variant tasks into the task/project graph may be difficult.

## Option 2) New `extends` field in tasks

Another option is to support an `extends` field within a task’s config, that points to another task (within the current project) to extend from.

The extending task config can define all task settings that the base task would (inputs, outputs, options, etc), except for `command`, which is replaced with `extends`. To modify arguments, use the `args` setting instead. Merge strategies will be applied when merging the base extended and extending tasks.

```yaml
tasks:
  build:
    command: 'cargo build'
    inputs:
      - 'crates/*/src/**/*'
      - 'crates/*/benches/**/*'
      - 'crates/*/tests/**/*'
      - 'crates/*/Cargo.toml'
      - 'Cargo.toml'
    outputs:
      - 'target/debug/moon'

  build-release:
    extends: 'build'
    args: '--release'
    outputs:
      - 'target/release/moon'
    options:
      mergeOutputs: 'replace'
      runInCI: false

  lint:
    command: 'eslint .'
    inputs:
      - '@globs(sources)'
      - '@globs(tests)'
      - '*.js'
      - '.eslintrc.js'
      - 'tsconfig.json'
      - '/.eslintignore'
      - '/.eslintrc.js'
      - '/tsconfig.eslint.json'
      - '/tsconfig.options.json'

  lint-fix:
    extends: 'lint'
    args: '--fix'
    local: true
```

### Pros

- YAML structure hasn’t changed. Much easier to read.
- Task inheritance will work as-is.
- Large settings (like `inputs`) can be entirely omitted for variant tasks.
- No additional learning curve.

### Cons

- Extending task _must come after_ the base task, otherwise an error will occur.
- No variant syntax for the target identifier.

## Option 3) Do nothing

And finally, we can simply do nothing as the current situation is working just fine.

## Option 4) Community suggestion

With that being said, we’d love to hear from the community if you have an idea that would work, and that is different from the options above!

**Comment by elmpp** [comment]
I would advocate for option 2, and actually looked for similar functionality myself just today with the `extend` semantic. 

It would be a logical extension of the existing `extends` at the different file levels

**Comment by theoephraim** [comment]
@milesj - there is actually a viable solution that currently works using yaml [anchors](https://www.educative.io/blog/advanced-yaml-syntax-cheatsheet#anchors)

An example:
```yaml
tasks:
  lint: &lint-base # base is reused below
    command: 'eslint src --ext .ts,.js,.cjs,.mjs,.vue'
    inputs:
      - '@group(sources)'
      - blah...
  lint/fix:
    <<: *lint-base # copies all the content in lint into here
    args: '--fix'
```

That said, while it works, it doesn't feel great and I think we can do better... I would vote for option 2.

**Comment by milesj** [comment]
Now available in 1.12: https://moonrepo.dev/blog/moon-v1.12
